### PR TITLE
[Merged by Bors] - feat: the equational criterion for flatness

### PR DESCRIPTION
**Pull request by trivial1711**
We prove the [equational criterion for flatness](https://stacks.math.columbia.edu/tag/00HK). As a consequence, every homomorphism from a finitely presented module to a flat module factors through a finite free module.

---

- [x] depends on: #12647

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/RingTheory/Flat/EquationalCriterion.lean [hunk 1]
```diff
@@ -0,0 +1,305 @@
...
+$a \circ f = 0$. -/
+theorem exists_factorization_of_comp_eq_zero_of_free [Flat R M] {K N : Type u}
+    [AddCommGroup K] [Module R K] [Finite R K] [AddCommGroup N] [Module R N] [Free R N] [Finite R N]
+    {f : K →ₗ[R] N} {x : N →ₗ[R] M} (h : x ∘ₗ f = 0) :
+    ∃ (κ : Type u) (_ : Fintype κ) (a : N →ₗ[R] (κ →₀ R)) (y : (κ →₀ R) →ₗ[R] M),
+      x = y ∘ₗ a ∧ a ∘ₗ f = 0 := by
+  have : ∀ (K' : Submodule R K) (_ : K'.FG), ∃ (κ : Type u) (_ : Fintype κ) (a : N →ₗ[R] (κ →₀ R))
+      (y : (κ →₀ R) →ₗ[R] M), x = y ∘ₗ a ∧ K' ≤ LinearMap.ker (a ∘ₗ f) := by
+    intro K' hK'
+    revert N
```
Comment (chrisflav) [inline Mathlib/RingTheory/Flat/EquationalCriterion.lean:274]:
```suggestion
  have (K' : Submodule R K) (hK' : K'.FG) : ∃ (κ : Type u) (_ : Fintype κ) (a : N →ₗ[R] (κ →₀ R))
      (y : (κ →₀ R) →ₗ[R] M), x = y ∘ₗ a ∧ K' ≤ LinearMap.ker (a ∘ₗ f) := by
    revert N
    apply Submodule.fg_induction (P := _) (N := K') (hN := hK')
```

File: Mathlib/RingTheory/Flat/EquationalCriterion.lean [hunk 2]
```diff
@@ -0,0 +1,305 @@
...
+      use κ₂, hκ₂, a₂ ∘ₗ a₁, y₂
+      simp_rw [comp_assoc]
+      exact ⟨trivial, sup_le (ha₁.trans (ker_le_ker_comp _ _)) ha₂⟩
+  convert this ⊤ Finite.out
+  simp only [top_le_iff, ker_eq_top]
+
+/-- Every homomorphism from a finitely presented module to a flat module factors through a finite
+free module. -/
+theorem exists_factorization_of_isFinitelyPresented [Flat R M] {P : Type u} [AddCommGroup P]
+    [Module R P] [fp : FinitePresentation R P] (h₁ : P →ₗ[R] M) :
```
Comment (chrisflav) [inline Mathlib/RingTheory/Flat/EquationalCriterion.lean:292]:
```suggestion
    [Module R P] [FinitePresentation R P] (h₁ : P →ₗ[R] M) :
```

File: Mathlib/RingTheory/Flat/EquationalCriterion.lean [hunk 3]
```diff
@@ -0,0 +1,305 @@
...
+      ∃ (κ : Type u) (_ : Fintype κ) (h₂ : P →ₗ[R] (κ →₀ R)) (h₃ : (κ →₀ R) →ₗ[R] M),
+        h₁ = h₃ ∘ₗ h₂ := by
+  obtain ⟨L, _, _, _, _, K, hK, ⟨ϕ⟩⟩ := FinitePresentation.equiv_quotient R P
+  haveI : Finite R ↥K := Module.Finite.iff_fg.mpr hK
+  have : (h₁ ∘ₗ ϕ.symm ∘ₗ K.mkQ) ∘ₗ K.subtype = 0 := by
+    simp_rw [comp_assoc, (LinearMap.exact_subtype_mkQ K).linearMap_comp_eq_zero, comp_zero]
+  obtain ⟨κ, hκ, a, y, ⟨hay, ha⟩⟩ := exists_factorization_of_comp_eq_zero_of_free this
```
Comment (chrisflav) [inline Mathlib/RingTheory/Flat/EquationalCriterion.lean:299]:
```suggestion
  obtain ⟨κ, hκ, a, y, hay, ha⟩ := exists_factorization_of_comp_eq_zero_of_free this
```

**Review by chrisflav — COMMENTED** [review]
That's a very nice result! I left some comments below.

File: Mathlib/RingTheory/Flat/EquationalCriterion.lean [hunk 4]
```diff
@@ -0,0 +1,305 @@
...
+open scoped BigOperators
+
+namespace Module
+
+variable {ι : Type u} [Fintype ι] (f : ι → R) (x : ι → M)
+
+/-- The proposition that the relation $\sum_i f_i x_i = 0$ in $M$ is trivial.
+That is, there exist a finite index type $\kappa$, elements
+$(y_j)_{j \in \kappa}$ of $M$, and elements $(a_{ij})_{i \in \iota, j \in \kappa}$ of $R$
+such that for all $i$,
```
[the remark this answers is not in the document]
Comment (trivial1711) [reply Mathlib/RingTheory/Flat/EquationalCriterion.lean:68]: It looks good in the doc preview

File: Mathlib/RingTheory/Flat/EquationalCriterion.lean [hunk 5]
```diff
@@ -0,0 +1,305 @@
...
+$x \colon R^{\iota} \to M$ such that $x(f) = 0$, there exist a finite index type $\kappa$ and module
+homomorphisms $a \colon R^{\iota} \to R^{\kappa}$ and $y \colon R^{\kappa} \to M$ such
+that $x = y \circ a$ and $a(f) = 0$.
+* For all finite free modules $N$, all elements $f \in N$, and all homomorphisms $x \colon N \to M$
+such that $x(f) = 0$, there exist a finite index type $\kappa$ and module homomorphisms
+$a \colon N \to R^{\kappa}$ and $y \colon R^{\kappa} \to M$ such that $x = y \circ a$ and
+$a(f) = 0$. -/
+theorem tfae_equational_criterion : List.TFAE [
+    Flat R M,
+    ∀ (I : Ideal R), Function.Injective ⇑(rTensor M (Submodule.subtype I)),
```
[the remark this answers is not in the document]
Comment (trivial1711) [reply Mathlib/RingTheory/Flat/EquationalCriterion.lean:117]: Yes, it makes the proof more convenient.

**Comment by trivial1711** [comment]
Hello, I managed to make `Algebra/Module/FinitePresentation.lean` fully universe polymorphic. This had the side effect of somehow breaking `RingTheory/Flat/EquationalCriterion.lean`, even though I didn't even touch that file. I think this is almost certainly not worth it. #13149

**Comment by chrisflav** [comment]
> Hello, I managed to make `Algebra/Module/FinitePresentation.lean` fully universe polymorphic. This had the side effect of somehow breaking `RingTheory/Flat/EquationalCriterion.lean`, even though I didn't even touch that file. I think this is almost certainly not worth it. #13149 

Oh great, thanks! I think the extra leg-freedom in the universe level of the indexing type will pay off, but I think we can merge this PR and generalize in a second step with your follow-up PR.

**Comment by trivial1711** [comment]
Hi, I'm sorry about the way I worded my comment earlier. I shouldn't be so dismissive about universe polymorphism; I now see it could very plausibly matter at some point. Do you have any suggestions for how to do #13149? You can make commits directly if you want.

**Comment by chrisflav** [comment]
> Hi, I'm sorry about the way I worded my comment earlier. I shouldn't be so dismissive about universe polymorphism; I now see it could very plausibly matter at some point. Do you have any suggestions for how to do #13149? You can make commits directly if you want.

Currently I am busy, but I'll have a look in two days.

File: Mathlib/RingTheory/Flat/EquationalCriterion.lean [hunk 6]
```diff
@@ -0,0 +1,305 @@
...
+      x f = 0 →
+        ∃ (κ : Type u) (_ : Fintype κ) (a : N →ₗ[R] (κ →₀ R)) (y : (κ →₀ R) →ₗ[R] M),
+          x = y ∘ₗ a ∧ a f = 0] := by
+  tfae_have 1 ↔ 2
+  · exact iff_rTensor_injective' R M
+  tfae_have 3 ↔ 2
+  · exact forall_vanishesTrivially_iff_forall_rTensor_injective R
+  tfae_have 3 ↔ 4
+  · simp [(TensorProduct.lid R M).injective.eq_iff.symm, isTrivialRelation_iff_vanishesTrivially]
+  tfae_have 4 → 5
```
[the remark this answers is not in the document]
Comment (chrisflav) [reply Mathlib/RingTheory/Flat/EquationalCriterion.lean:134]: Whats your stance on this?

File: Mathlib/Algebra/Module/FinitePresentation.lean [hunk 7]
```diff
@@ -69,6 +69,20 @@ section Ring
...
 variable (R M N) [Ring R] [AddCommGroup M] [Module R M] [AddCommGroup N] [Module R N]
 
+universe u in
+variable (R M : Type u) [Ring R] [AddCommGroup M] [Module R M] in
+/-- A finitely presented module is isomorphic to the quotient of a finite free module by a finitely
+generated submodule. -/
+theorem Module.FinitePresentation.equiv_quotient [fp : Module.FinitePresentation R M] :
+    ∃ (L : Type u) (_ : AddCommGroup L) (_ : Module R L) (_ : Module.Free R L)
+      (_ : Module.Finite R L) (K : Submodule R L) (_ : K.FG), Nonempty (M ≃ₗ[R] L ⧸ K) := by
+  obtain ⟨ι, ⟨hι₁, hι₂⟩⟩ := fp
```
[the remark this answers is not in the document]
Comment (chrisflav) [reply Mathlib/Algebra/Module/FinitePresentation.lean]: Postponed to #13149.

File: Mathlib/RingTheory/Flat/EquationalCriterion.lean [hunk 8]
```diff
@@ -0,0 +1,305 @@
...
+    ∀ {ι : Type u} [Fintype ι] {f : ι → R} {x : ι → M}, ∑ i, f i ⊗ₜ x i = (0 : R ⊗[R] M) →
+      VanishesTrivially R f x,
+    ∀ {ι : Type u} [Fintype ι] {f : ι → R} {x : ι → M}, ∑ i, f i • x i = 0 → IsTrivialRelation f x,
+    ∀ {ι : Type u} [Fintype ι] {f : ι →₀ R} {x : (ι →₀ R) →ₗ[R] M}, x f = 0 →
+      ∃ (κ : Type u) (_ : Fintype κ) (a : (ι →₀ R) →ₗ[R] (κ →₀ R)) (y : (κ →₀ R) →ₗ[R] M),
+        x = y ∘ₗ a ∧ a f = 0,
+    ∀ {N : Type u} [AddCommGroup N] [Module R N] [Free R N] [Finite R N] {f : N} {x : N →ₗ[R] M},
+      x f = 0 →
+        ∃ (κ : Type u) (_ : Fintype κ) (a : N →ₗ[R] (κ →₀ R)) (y : (κ →₀ R) →ₗ[R] M),
+          x = y ∘ₗ a ∧ a f = 0] := by
```
[the remark this answers is not in the document]
Comment (chrisflav) [reply Mathlib/RingTheory/Flat/EquationalCriterion.lean:127]: Also postponed to #13149 I think.

File: Mathlib/RingTheory/Flat/EquationalCriterion.lean [hunk 9]
```diff
@@ -0,0 +1,305 @@
...
+
+variable {R M : Type u} [CommRing R] [AddCommGroup M] [Module R M]
+
+open Classical DirectSum LinearMap TensorProduct Finsupp
+open scoped BigOperators
+
+namespace Module
+
+variable {ι : Type u} [Fintype ι] (f : ι → R) (x : ι → M)
+
```
[the remark this answers is not in the document]
Comment (chrisflav) [reply Mathlib/RingTheory/Flat/EquationalCriterion.lean:64]: Postponed to #13149 .

**Comment by chrisflav** [comment]
Apart from one open question above, I think this is good to go. Since the universe generalizations depend on #13149, we can merge this one as is and generalize after #13149.

**Review by jcommelin — COMMENTED** [review]
Thanks :tada:

bors merge

**Pull request closed without merging**